### ISSUE TITLE
Better union error

### ISSUE THREAD

**Issue by tinganho** [microsoft/TypeScript#18403]
I think TS can improve a lot by showing the errors on unions correctly.

It would be good if it could pick the type in the union where it matches the most in cases where no types matches. 

In the below case, we at least match on a property name, but not property value.

My real world case took a hours to figure out what was wrong:

```ts
interface ViewData {
    view: string;
    data?: number;
}

interface ViewDataArgument {
    [region: string]: ViewData | ViewData[];
}

let decl: ViewDataArgument = {
    'r': { // Property 'r' is not assignable to 'ViewData[]'
        view: 1, // What really went wrong, was that I assigned a number to a string.
    }
}
```
http://www.typescriptlang.org/play/index.html#src=interface%20ViewData%20%7B%0A%20%20%20%20view%3A%20string%3B%0A%20%20%20%20data%3F%3A%20number%3B%0A%7D%0A%0Ainterface%20ViewDataArgument%20%7B%0A%20%20%20%20%5Bregion%3A%20string%5D%3A%20ViewData%20%7C%20ViewData%5B%5D%3B%0A%7D%0A%0Alet%20decl%3A%20ViewDataArgument%20%3D%20%7B%0A%20%20%20%20'r'%3A%20%7B%20%2F%2F%20Property%20'r'%20is%20not%20assignable%20to%20'ViewData%5B%5D'%0A%20%20%20%20%20%20%20%20view%3A%201%2C%20%2F%2F%20What%20really%20was%20wrong%20is%20that%20I%20assigned%20a%20number%20to%20a%20string.%0A%20%20%20%20%7D%0A%7D

**Comment by RyanCavanaugh** [comment]
This reports the expected error now 🙂
